Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm

```diff
@@ -233,21 +279,29 @@
     return details;
 }
 
-void WebExtensionContextProxy::resourceLoadDidSendRequest(WebExtensionTabIdentifier tabID, WebExtensionWindowIdentifier windowID, const WebCore::ResourceRequest& request, const ResourceLoadInfo& resourceLoad)
+void WebExtensionContextProxy::resourceLoadDidSendRequest(WebExtensionTabIdentifier tabID, WebExtensionWindowIdentifier windowID, const WebCore::ResourceRequest& request, const ResourceLoadInfo& resourceLoad, const std::optional<IPC::FormDataReference>& formDataOptional)
 {
     auto *details = webRequestDetailsForResourceLoad(resourceLoad, tabID);
 
     // FIXME: <rdar://problem/59922101> Chrome documentation says this about requestBody:
     // Only provided if extraInfoSpec contains 'requestBody'.
-    if (auto *requestBody = request.nsURLRequest(WebCore::HTTPBodyUpdatePolicy::DoNotUpdateHTTPBody).HTTPBody)
-        details[@"requestBody"] = convertRequestBodyToWebExtensionFormat(requestBody);
+
+    RefPtr formData = formDataOptional ? formDataOptional->data() : nullptr;
+    auto contentType = request.httpContentType();
 
     enumerateNamespaceObjects([&](auto& namespaceObject) {
-        auto& webRequestObject = namespaceObject.webRequest();
-        webRequestObject.onBeforeRequest().invokeListenersWithArgument(details, tabID, windowID, resourceLoad);
+        namespaceObject.webRequest().onBeforeRequest().enumerateListeners(tabID, windowID, resourceLoad, [details = RetainPtr { details }, formData, contentType](WebExtensionCallbackHandler& listener) {
+            if (formData) {
+                if (auto *requestBody = toWebAPI(*formData, contentType, listener.globalContext()))
+                    [details setObject:requestBody forKey:@"requestBody"];
+            }
+
+            listener.call(details.get());
+
+            [details removeObjectForKey:@"requestBody"];
```

Comment:
Nit: Should `requestBody` be in a constant somewhere?

Reply:
We didn't use constants for some reaosn in this file, so I kept with that. We should consider converting all the strings in here to constants. I have another PR coming that touches webRequest, so I will do that in a follow up to avoid merge issues.